contrib/lib/src/templates/fairing.rs

```diff
@@ -0,0 +1,154 @@
+
+    use super::{Context, Engines};
+
+    use self::notify::{raw_watcher, RawEvent, RecommendedWatcher, RecursiveMode, Watcher};
+
+    /// Wraps a Context. With `cfg(debug_assertions)` active, this structure
+    /// additionally provides a method to reload the context at runtime.
+    pub struct ContextManager {
+        /// The current template context, inside an RwLock so it can be updated
+        context: RwLock<Context>,
+        /// A filesystem watcher and the receive queue for its events
+        watcher: Option<(RecommendedWatcher, Mutex<Receiver<RawEvent>>)>,
+    }
+
+    impl ContextManager {
+        pub fn new(ctxt: Context) -> ContextManager {
+            let (tx, rx) = channel();
+
+            let watcher = if let Ok(mut watcher) = raw_watcher(tx) {
+                if watcher.watch(ctxt.root.clone(), RecursiveMode::Recursive).is_ok() {
+                    Some((watcher, Mutex::new(rx)))
+                } else {
+                    warn!("Could not monitor the templates directory for changes.");
+                    warn!("Live template reload will be unavailable");
+                    None
+                }
+            } else {
+                warn!("Could not instantiate a filesystem watcher.");
+                warn!("Live template reload will be unavailable");
+                None
+            };
+
+            ContextManager {
+                watcher,
+                context: RwLock::new(ctxt),
+            }
+        }
+
+        pub fn get<'a>(&'a self) -> impl Deref<Target=Context> + 'a {
+            self.context.read().unwrap()
+        }
+
+        fn get_mut<'a>(&'a self) -> impl DerefMut<Target=Context> + 'a {
+            self.context.write().unwrap()
+        }
+
+        pub fn reload_if_needed<F: Fn(&mut Engines)>(&self, custom_callback: F) {
+            self.watcher.as_ref().map(|w| {
+                let rx = w.1.lock().expect("receive queue");
+                let mut changed = false;
+                while let Ok(_) = rx.try_recv() {
+                    changed = true;
+                }
+
+                if changed {
+                    warn!("Change detected, reloading templates");
+                    let mut ctxt = self.get_mut();
+                    if let Some(mut new_ctxt) = Context::initialize(ctxt.root.clone()) {
+                        custom_callback(&mut new_ctxt.engines);
+                        *ctxt = new_ctxt;
+                    } else {
+                        warn!("An error occurred while reloading templates.");
+                        warn!("The previous templates will remain active.");
+                    };
+                }
+            });
+        }
+    }
+}
+
+pub use self::context::ContextManager;
+
+pub struct TemplateFairing {
+    custom_callback: Box<Fn(&mut Engines) + Send + Sync + 'static>,
+}
+
+impl TemplateFairing {
+    pub fn new(custom_callback: Box<Fn(&mut Engines) + Send + Sync + 'static>) -> TemplateFairing {
```

Comment:
It looks like this is called only once. Why not write `TemplateFairing { custom_callback }` where this is called instead of having this method? That would be just as clear (if not clearer due to the heavy type signature here) while also being fewer lines of code.